File: app/src/main/java/com/example/kotlin_movieapp/adapters/MovieAdapter.java
package com.example.kotlin_movieapp.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.kotlin_movieapp.R;
import com.example.kotlin_movieapp.models.Movie;
import com.example.kotlin_movieapp.models.MovieSource;

public class MovieAdapter extends RecyclerView.Adapter <MovieAdapter.ViewHolder> {

    private final MovieSource movieSource;

    public MovieAdapter(MovieSource movieSource){
        this.movieSource = movieSource;
    }

    @NonNull
    @Override
    public MovieAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View itemView = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.movie_item, parent, false);

        return new MovieAdapter.ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.setData(movieSource.getMovie(position));
    }

    @Override
    public int getItemCount() {
        return movieSource.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        private final ImageView poster;
        private final TextView tvMovieName;

        public ViewHolder(View itemView) {
            super(itemView);

            poster = itemView.findViewById(R.id.moviePoster);
            tvMovieName = itemView.findViewById(R.id.movieName);

        }

        public void setData(Movie movie) {
            poster.setImageResource(movie.getImage());
            tvMovieName.setText(movie.getName());
        }
    }
}

File: app/src/main/java/com/example/kotlin_movieapp/adapters/MovieAdapter.kt
package com.example.kotlin_movieapp.adapters

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.example.kotlin_movieapp.ui.main.MainActivity
import com.example.kotlin_movieapp.R
import com.example.kotlin_movieapp.databinding.MovieItemBinding
import com.example.kotlin_movieapp.model.collectionResponse.CollectionItem
import com.example.kotlin_movieapp.ui.main.movieDetails.MovieDetailsFragment
import com.example.kotlin_movieapp.utils.KEY_BUNDLE_MOVIE
import com.squareup.picasso.Picasso

class MovieAdapter(
    private var movieData: List<CollectionItem>
    ) : RecyclerView.Adapter<MovieAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = MovieItemBinding.inflate(LayoutInflater.from(parent.context),
            parent,
            false)

        return ViewHolder(binding.root)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(movieData[position])
    }

    override fun getItemCount(): Int = movieData.size


    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun bind(movie: CollectionItem) {
            val binding = MovieItemBinding.bind(itemView)
            with(binding) {

                Picasso.get()?.load(movie.poster?.previewUrl)?.into(moviePoster)

                root.setOnClickListener {
                    (itemView.context as MainActivity).supportFragmentManager
                        .beginTransaction()
                        .addToBackStack("tag")
                        .add(R.id.main_container,
                            MovieDetailsFragment.newInstance(Bundle().apply {
                                putParcelable(KEY_BUNDLE_MOVIE, movie)
                            }))
                        .commit()
                }
            }
        }
    }
}